index.bs

```diff
@@ -4969,162 +4911,189 @@ The [=map/values=] of the [=context menu map=], which are
+      1. For each |urlPattern| in |properties|["{{
+          ContextMenusProperties/targetURLPatterns}}"]:
 
-  1. Let |callbackList| be the [=ContextMenusEvent/event listener list=] of [=this=].
+          1. If |targetUrl| [=matches a URLPattern=], given |urlPattern|,
+              set |matchesTargetUrl| to `true` and break;
 
-  1. [=list/For each=] |listItem| in |callbackList|:
+      1. If |matchesTargetUrl| is `false`, return `false`.
 
-    1. If |listItem| equals to |callback|, return `true`.
+  Note: Different user agents may implement additional conditions.
 
-  1. Return `false`.
+  1. Return `true`.
 
 </div>
 
-<div algorithm>
-
-  The <dfn method for=ContextMenusEvent>hasListeners()</dfn> method steps are:
-
-  1. Let |callbackList| be the [=ContextMenusEvent/event listener list=] of [=this=].
-
-  1. If |callbackList| [=list/is empty=], return `false`.
+## Events ## {#api-context-menus-events}
 
-  1. Otherwise, return `true`.
+{{ContextMenus}} fires the following events:
 
-</div>
-
-{{ContextMenus}} has the attribute {{ContextMenus/onShow}}, which is a
-{{ContextMenusEvent}} that is associated with [=/contextmenu=] [=events=].
-
-<div algorithm='onshow-event'>
+<table>
+  <tr>
+    <th>Event name</th>
+    <th>Interface</th>
+    <th>Fired when...</th>
+  </tr>
+  <tr>
+    <td><dfn for=ContextMenus>show</dfn></td>
+    <td>{{Event}}</td>
+    <td>a context menu is displayed to the user.</td>
+  </tr>
+  <tr>
+    <td><dfn for=ContextMenus>click</dfn></td>
+    <td>{{ContextMenusClickEvent}}</td>
+    <td>a [=context menu item=] is selected by the user.</td>
+  </tr>
+</table>
 
-To [= dispatch =] an {{ContextMenus/onShow}} [= event =] |event|, given {{HTMLControlledFrameElement}} |controlledFrame|:
+{{ContextMenus}} supports the following [=event handlers=] (and their
+corresponding [=event handler event types=]) as [=event handler IDL attributes=]:
 
-1. Let |callbackList| be the [=ContextMenusEvent/event listener list=] of |event|.
+<table>
+  <tr>
+    <th>[=Event handlers=]</th>
+    <th>[=Event handler event types=]</th>
+  </tr>
+  <tr>
+    <td><dfn attribute for=ContextMenus>onshow</dfn></td>
+    <td>[=ContextMenus/show=]</td>
+  </tr>
+  <tr>
+    <td><dfn attribute for=ContextMenus>onclick</dfn></td>
+    <td>[=ContextMenus/click=]</td>
+  </tr>
+</table>
 
-1. [=list/For each=] |callback| in |callbackList|:
+When one or more of a {{ContextMenus}} |contextMenus|'s [=ContextMenus/
+context menu items=] are shown to the user, [=ContextMenus/dispatch a
+show event=] to |contextMenus|.
 
-  1. [=Queue a global task=] on the [=DOM manipulation task source=] of |controlledFrame|'s [=relevant global object=] to invoke |callback|.
+When a [=ContextMenus/context menu item=] |item| is selected by a user,
+[=ContextMenus/dispatch a click event=] given the {{ContextMenus}} |item|
+belongs to, the {{HTMLElement}} inside the embedding document where the
```

Comment:
Can you help me understand what document this `{{HTMLElement}}` is a part of? Is the idea that the user clicks an audio/video/img/anchor element _inside_ the controlled frame's inner Document, and then magically the "determining whether the context menu item will be shown" algorithm is invoked with the controlled frame that hosts the clicked element, the element inside the frame, and .... how do we determine which `ContextMenuProperties` object is passed into this algorithm?

I guess I'm confused about the intended flow.